### Imports

In [99]:
%pip install pandas
import pandas as pd 
import os
from sqlalchemy import create_engine
from dotenv import load_dotenv, find_dotenv
from pathlib import Path


Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 25.3 -> 26.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [100]:
# connect to database
env_path = Path(__file__).parent / ".env" if "__file__" in dir() else Path("../.env")
load_dotenv(dotenv_path=env_path, override=True)

engine = create_engine(
    f"postgresql://{os.getenv('DB_USER')}:{os.getenv('DB_PASSWORD')}"
    f"@{os.getenv('DB_HOST_LOCAL')}:{os.getenv('DB_PORT_LOCAL')}/{os.getenv('DB_NAME')}"
)

### 1. Election Results Datasets

In [101]:
# read and initialise tables
df2004 = pd.read_sql("SELECT * FROM bronze.\"2004_npe\"", con=engine)
df2009 = pd.read_sql("SELECT * FROM bronze.\"2009_npe\"", con=engine)
df2014 = pd.read_sql("SELECT * FROM bronze.\"2014_npe\"", con=engine)
df2019 = pd.read_sql("SELECT * FROM bronze.\"2019_npe\"", con=engine)
df2024 = pd.read_sql("SELECT * FROM bronze.\"2024_npe\"", con=engine)


#### a. Naming conventions

In [102]:
df2004.head()

,ELECTORAL EVENT,PROVINCE,MUNICIPALITY,VOTING\nDISTRICT,PARTY NAME,REGISTERED\nVOTERS,% VOTER\nTURNOUT,VALID\nVOTES,SPOILT\nVOTES,TOTAL VOTES\nCAST
0,14 APR 2004 NATIONAL ELECTION,EASTERN CAPE,NaN,10590006,AFRICAN CHRISTIAN DEMOCRATIC PARTY,0,0.00%,0,1,9
1,14 APR 2004 NATIONAL ELECTION,EASTERN CAPE,NaN,10590006,AFRICAN NATIONAL CONGRESS,0,0.00%,0,1,9
2,14 APR 2004 NATIONAL ELECTION,EASTERN CAPE,NaN,10590006,AZANIAN PEOPLE'S ORGANISATION,0,0.00%,0,1,9
3,14 APR 2004 NATIONAL ELECTION,EASTERN CAPE,NaN,10590006,CHRISTIAN DEMOCRATIC PARTY,0,0.00%,0,1,9
4,14 APR 2004 NATIONAL ELECTION,EASTERN CAPE,NaN,10590006,DEMOCRATIC ALLIANCE/DEMOKRATIESE ALLIANSIE,0,0.00%,7,1,9


In [148]:
df2004.shape

(612135, 10)

In [103]:
df2004.columns

Index(['ELECTORAL EVENT', 'PROVINCE', 'MUNICIPALITY', 'VOTING\nDISTRICT',
       'PARTY NAME', 'REGISTERED\nVOTERS', '% VOTER\nTURNOUT', 'VALID\nVOTES',
       'SPOILT\nVOTES', 'TOTAL VOTES\nCAST'],
      dtype='str')

In [104]:
df2004_renamed=df2004.rename(columns={'ELECTORAL EVENT': 'electoral_event',
                       'PROVINCE': 'province',
                        'MUNICIPALITY': 'municipality',
                        'VOTING\nDISTRICT': 'vd_number',
                        'PARTY NAME': 'party_name',
                        'REGISTERED\nVOTERS': 'registered_population',
                        '% VOTER\nTURNOUT': 'voter_turnout_percent',
                        'VALID\nVOTES': 'valid_votes',
                        'SPOILT\nVOTES': 'spoilt_votes',
                        'TOTAL VOTES\nCAST': 'total_votes_cast'})

In [105]:
df2009.head()

,ELECTORAL EVENT,PROVINCE,MUNICIPALITY,WARD,VOTING\nDISTRICT,PARTY NAME,REGISTERED\nVOTERS,% VOTER\nTURNOUT,VALID\nVOTES,SPOILT\nVOTES,TOTAL VOTES\nCAST,SECTION 24A\nVOTES,SPECIAL\nVOTES
0,22 APR 2009 NATIONAL ELECTION,EASTERN CAPE,EASTERN CAPE - OUT OF COUNTRY,NaN,10590006,A PARTY,NaN,NaN,0,9,620,0,620
1,22 APR 2009 NATIONAL ELECTION,EASTERN CAPE,EASTERN CAPE - OUT OF COUNTRY,NaN,10590006,AFRICAN CHRISTIAN DEMOCRATIC PARTY,NaN,NaN,9,9,620,0,620
2,22 APR 2009 NATIONAL ELECTION,EASTERN CAPE,EASTERN CAPE - OUT OF COUNTRY,NaN,10590006,AFRICAN NATIONAL CONGRESS,NaN,NaN,38,9,620,0,620
3,22 APR 2009 NATIONAL ELECTION,EASTERN CAPE,EASTERN CAPE - OUT OF COUNTRY,NaN,10590006,AFRICAN PEOPLE'S CONVENTION,NaN,NaN,0,9,620,0,620
4,22 APR 2009 NATIONAL ELECTION,EASTERN CAPE,EASTERN CAPE - OUT OF COUNTRY,NaN,10590006,AL JAMA-AH,NaN,NaN,0,9,620,0,620


In [106]:
df2009.columns

Index(['ELECTORAL EVENT', 'PROVINCE', 'MUNICIPALITY', 'WARD',
       'VOTING\nDISTRICT', 'PARTY NAME', 'REGISTERED\nVOTERS',
       '% VOTER\nTURNOUT', 'VALID\nVOTES', 'SPOILT\nVOTES',
       'TOTAL VOTES\nCAST', 'SECTION 24A\nVOTES', 'SPECIAL\nVOTES'],
      dtype='str')

In [107]:
df2009_renamed=df2009.rename(columns={'ELECTORAL EVENT': 'electoral_event',
                       'PROVINCE': 'province',
                        'MUNICIPALITY': 'municipality',
                        'WARD': 'ward_number',
                        'VOTING\nDISTRICT': 'vd_number',
                        'PARTY NAME': 'party_name',
                        'REGISTERED\nVOTERS': 'registered_population',
                        '% VOTER\nTURNOUT': 'voter_turnout_percent',
                        'VALID\nVOTES': 'valid_votes',
                        'SPOILT\nVOTES': 'spoilt_votes',
                        'TOTAL VOTES\nCAST': 'total_votes_cast',
                        'SECTION 24A\nVOTES': 'section24a_votes',
                        'SPECIAL\nVOTES': 'special_votes'})

In [108]:
df2014.head()

,ELECTORAL EVENT,PROVINCE,MUNICIPALITY,WARD,VOTING DISTRICT,PARTY NAME,REGISTERED VOTERS,% VOTER TURNOUT,VALID VOTES,SPOILT VOTES,TOTAL VOTES CAST,SECTION 24A VOTES,SPECIAL VOTES
0,2014 NATIONAL ELECTION,EASTERN CAPE,BUF - BUFFALO CITY METROPOLITAN MUNICIPALITY [...,29200001.0,10590151,AFRICAN CHRISTIAN DEMOCRATIC PARTY,2316,0.589052,6,24,1442,132,10
1,2014 NATIONAL ELECTION,EASTERN CAPE,BUF - BUFFALO CITY METROPOLITAN MUNICIPALITY [...,29200001.0,10590151,AFRICAN INDEPENDENT CONGRESS,2316,0.589052,7,24,1442,132,10
2,2014 NATIONAL ELECTION,EASTERN CAPE,BUF - BUFFALO CITY METROPOLITAN MUNICIPALITY [...,29200001.0,10590151,AFRICAN NATIONAL CONGRESS,2316,0.589052,505,24,1442,132,10
3,2014 NATIONAL ELECTION,EASTERN CAPE,BUF - BUFFALO CITY METROPOLITAN MUNICIPALITY [...,29200001.0,10590151,AFRICAN PEOPLE'S CONVENTION,2316,0.589052,2,24,1442,132,10
4,2014 NATIONAL ELECTION,EASTERN CAPE,BUF - BUFFALO CITY METROPOLITAN MUNICIPALITY [...,29200001.0,10590151,AGANG SOUTH AFRICA,2316,0.589052,0,24,1442,132,10


In [109]:
df2014.columns

Index(['ELECTORAL EVENT', 'PROVINCE', 'MUNICIPALITY', 'WARD',
       'VOTING DISTRICT', 'PARTY NAME', 'REGISTERED VOTERS', '% VOTER TURNOUT',
       'VALID VOTES', 'SPOILT VOTES', 'TOTAL VOTES CAST', 'SECTION 24A VOTES',
       'SPECIAL VOTES'],
      dtype='str')

In [110]:
df2014_renamed=df2014.rename(columns={'ELECTORAL EVENT': 'electoral_event',
                       'PROVINCE': 'province',
                        'MUNICIPALITY': 'municipality',
                        'WARD': 'ward_number',
                        'VOTING DISTRICT': 'vd_number',
                        'PARTY NAME': 'party_name',
                        'REGISTERED VOTERS': 'registered_population',
                        '% VOTER TURNOUT': 'voter_turnout_percent',
                        'VALID VOTES': 'valid_votes',
                        'SPOILT VOTES': 'spoilt_votes',
                        'TOTAL VOTES CAST': 'total_votes_cast',
                        'SECTION 24A VOTES': 'section24a_votes',
                        'SPECIAL VOTES': 'special_votes'})

In [111]:
df2014_renamed.head()

,electoral_event,province,municipality,ward_number,vd_number,party_name,registered_population,voter_turnout_percent,valid_votes,spoilt_votes,total_votes_cast,section24a_votes,special_votes
0,2014 NATIONAL ELECTION,EASTERN CAPE,BUF - BUFFALO CITY METROPOLITAN MUNICIPALITY [...,29200001.0,10590151,AFRICAN CHRISTIAN DEMOCRATIC PARTY,2316,0.589052,6,24,1442,132,10
1,2014 NATIONAL ELECTION,EASTERN CAPE,BUF - BUFFALO CITY METROPOLITAN MUNICIPALITY [...,29200001.0,10590151,AFRICAN INDEPENDENT CONGRESS,2316,0.589052,7,24,1442,132,10
2,2014 NATIONAL ELECTION,EASTERN CAPE,BUF - BUFFALO CITY METROPOLITAN MUNICIPALITY [...,29200001.0,10590151,AFRICAN NATIONAL CONGRESS,2316,0.589052,505,24,1442,132,10
3,2014 NATIONAL ELECTION,EASTERN CAPE,BUF - BUFFALO CITY METROPOLITAN MUNICIPALITY [...,29200001.0,10590151,AFRICAN PEOPLE'S CONVENTION,2316,0.589052,2,24,1442,132,10
4,2014 NATIONAL ELECTION,EASTERN CAPE,BUF - BUFFALO CITY METROPOLITAN MUNICIPALITY [...,29200001.0,10590151,AGANG SOUTH AFRICA,2316,0.589052,0,24,1442,132,10


In [112]:
df2019.head()


,Province,Municipality,VD Number,VS Name,Registered Population,Spoilt Votes,Total Valid Votes,sPartName,Party Votes,Generated Datetime: 30 Jun 2020 13:27:18
0,Eastern Cape,BUF - Buffalo City,10590016,NU 12 COMMUNITY HALL,2791.0,15,1689,AFRICAN CHRISTIAN DEMOCRATIC PARTY,6,None
1,Eastern Cape,BUF - Buffalo City,10590016,NU 12 COMMUNITY HALL,2791.0,15,1689,AFRICAN CONGRESS OF DEMOCRATS,0,None
2,Eastern Cape,BUF - Buffalo City,10590016,NU 12 COMMUNITY HALL,2791.0,15,1689,AFRICAN CONTENT MOVEMENT,0,None
3,Eastern Cape,BUF - Buffalo City,10590016,NU 12 COMMUNITY HALL,2791.0,15,1689,AFRICAN COVENANT,0,None
4,Eastern Cape,BUF - Buffalo City,10590016,NU 12 COMMUNITY HALL,2791.0,15,1689,AFRICAN DEMOCRATIC CHANGE,0,None


In [113]:
df2019.columns

Index(['Province', 'Municipality', 'VD Number', 'VS Name',
       'Registered Population', 'Spoilt Votes', 'Total Valid Votes',
       'sPartName', 'Party Votes', 'Generated Datetime: 30 Jun 2020 13:27:18'],
      dtype='str')

In [114]:
df2019_renamed=df2019.rename(columns={'Province': 'province',
                        'Municipality': 'municipality',
                        'VD Number': 'vd_number',
                        'VS Name': 'vs_name', 
                        'sPartName': 'party_name',
                        'Registered Population': 'registered_population',
                        'Party Votes': 'party_votes',
                        'Total Valid Votes': 'valid_votes',
                        'Spoilt Votes': 'spoilt_votes'})

In [115]:
df2024.head()

,Province,Municipality,VD_Number,VS_Name,Registered_Population,Spoilt_Votes,Total_Valid_Votes,sPartyName,Party_Votes,Generated_Datetime
0,Eastern Cape,BUF - Buffalo City,10590016,NU 12 COMMUNITY HALL,2622,25,1538,#HOPE4SA,0,06 Jun 2024 03:28:34 PM
1,Eastern Cape,BUF - Buffalo City,10590016,NU 12 COMMUNITY HALL,2622,25,1538,ABANTU BATHO CONGRESS,1,06 Jun 2024 03:28:34 PM
2,Eastern Cape,BUF - Buffalo City,10590016,NU 12 COMMUNITY HALL,2622,25,1538,ABLE LEADERSHIP,0,06 Jun 2024 03:28:34 PM
3,Eastern Cape,BUF - Buffalo City,10590016,NU 12 COMMUNITY HALL,2622,25,1538,ACTION ALLIANCE DEVELOPMENT PARTY,0,06 Jun 2024 03:28:34 PM
4,Eastern Cape,BUF - Buffalo City,10590016,NU 12 COMMUNITY HALL,2622,25,1538,ACTIONSA,9,06 Jun 2024 03:28:34 PM


In [116]:
df2024.columns

Index(['Province', 'Municipality', 'VD_Number', 'VS_Name',
       'Registered_Population', 'Spoilt_Votes', 'Total_Valid_Votes',
       'sPartyName', 'Party_Votes', 'Generated_Datetime'],
      dtype='str')

In [117]:
df2024_renamed=df2024.rename(columns={'Province': 'province',
                    'Municipality': 'municipality',
                    'VD_Number': 'vd_number',
                    'VS_Name': 'vs_name',
                    'Registered_Population': 'registered_population',
                    'Spoilt_Votes': 'spoilt_votes',
                    'Total_Valid_Votes': 'valid_votes',
                    'sPartyName': 'party_name',
                    'Generated_Datetime': 'generated_datetime',
                    'Party_Votes': 'party_votes'})


#### b. Missing Values

In [118]:
print(df2004_renamed.isnull().sum())

electoral_event            0
province                   0
municipality             189
vd_number                  0
party_name                 0
registered_population      0
voter_turnout_percent      0
valid_votes                0
spoilt_votes               0
total_votes_cast           0
dtype: int64


The missing values in the "Municipality"column will be filled in by the values found in the rows with the same district number, as a ditrict number will always be under the same municipality.

In [119]:
# lookup from rows that DO have a municipality
municipality_lookup = (
    df2004_renamed[df2004_renamed['municipality'].notna()]
    .groupby('vd_number')['municipality']
    .first()
)

# Fill missing using lookup, otherwise, replace with 'unknown'
df2004_renamed['municipality'] = df2004_renamed['municipality'].fillna(
    df2004_renamed['vd_number'].map(municipality_lookup)
).fillna('UNKNOWN')


print(f"Remaining missing: {df2004_renamed['municipality'].isna().sum()}")

Remaining missing: 0


In [120]:
print(df2014_renamed.isnull().sum())

electoral_event           0
province                  0
municipality             29
ward_number              29
vd_number                 0
party_name                0
registered_population     0
voter_turnout_percent     0
valid_votes               0
spoilt_votes              0
total_votes_cast          0
section24a_votes          0
special_votes             0
dtype: int64


As before, we will fill in the missing values by creating a lookup table from rows where the municipality and ward number are known, mapping them to their respective voting districts. Any voting districts with no municipality or ward number on any row will be set to 'UNKNOWN'.

In [121]:
# Fill municipality
municipality_lookup = (
    df2009_renamed[df2009_renamed['municipality'].notna()]
    .groupby('vd_number')['municipality']
    .first()
)
df2009_renamed['municipality'] = df2009_renamed['municipality'].fillna(
    df2009_renamed['vd_number'].map(municipality_lookup)
).fillna('UNKNOWN')

# Fill ward_number
ward_lookup = (
    df2009_renamed[df2009_renamed['ward_number'].notna()]
    .groupby('vd_number')['ward_number']
    .first()
)
df2009_renamed['ward_number'] = df2009_renamed['ward_number'].fillna(
    df2009_renamed['vd_number'].map(ward_lookup)
).fillna('UNKNOWN')


print(df2009_renamed[['municipality', 'ward_number']].isna().sum())

municipality    0
ward_number     0
dtype: int64


In [122]:
print(df2019_renamed.isnull().sum())

province                                          0
municipality                                      0
vd_number                                         0
vs_name                                           0
registered_population                           125
spoilt_votes                                      0
valid_votes                                       0
party_name                                      126
party_votes                                       0
Generated Datetime: 30 Jun 2020 13:27:18    1099566
dtype: int64


The "party_name" column has 126 missing values which will be replaced with 'UNKNOWN' for traceability. The "Generated Datetime" column is IEC export metadata with over 1 million missing values and no analytical value, so it will be dropped entirely.

In [123]:
df2019_renamed['party_name'] = df2019_renamed['party_name'].fillna('UNKNOWN')

# Drop the generated datetime column entirely - no analytical value
df2019_renamed = df2019_renamed.drop(columns=['Generated Datetime: 30 Jun 2020 13:27:18'])

print(df2019_renamed[['party_name']].isna().sum())

party_name    0
dtype: int64


#### c. Data Types

In [124]:
# Check datatypes for all years
for year, df in {'2004': df2004_renamed, '2009': df2009_renamed, '2014': df2014_renamed, '2019': df2019_renamed, '2024': df2024_renamed}.items():
    print(f"\n{'='*50}")
    print(f" {year} NPE")
    print(f"{'='*50}")
    print(df.dtypes)


 2004 NPE
electoral_event            str
province                   str
municipality               str
vd_number                int64
party_name                 str
registered_population    int64
voter_turnout_percent      str
valid_votes              int64
spoilt_votes             int64
total_votes_cast         int64
dtype: object

 2009 NPE
electoral_event             str
province                    str
municipality                str
ward_number              object
vd_number                 int64
party_name                  str
registered_population       str
voter_turnout_percent       str
valid_votes                 str
spoilt_votes              int64
total_votes_cast            str
section24a_votes            str
special_votes               str
dtype: object

 2014 NPE
electoral_event              str
province                     str
municipality                 str
ward_number              float64
vd_number                  int64
party_name                   str
registered_popu

We will not fix the inconsistent datatypes in the election results datasets.

In [125]:
df2004_renamed.dtypes

electoral_event            str
province                   str
municipality               str
vd_number                int64
party_name                 str
registered_population    int64
voter_turnout_percent      str
valid_votes              int64
spoilt_votes             int64
total_votes_cast         int64
dtype: object

In [126]:
# Remove % symbol and convert to float
df2004_renamed['voter_turnout_percent'] = (
    df2004_renamed['voter_turnout_percent']
    .str.replace('%', '', regex=False)
    .astype(float)
)

df2009_renamed['voter_turnout_percent'] = (
    df2009_renamed['voter_turnout_percent']
    .str.replace('%', '', regex=False)
    .astype(float)
)


# Convert ward_number to nullable integer
df2009_renamed['ward_number'] = pd.to_numeric(df2009_renamed['ward_number'], errors='coerce').astype('Int64')

# convert rest if string columns to int
int_cols = [
    'registered_population',
    'valid_votes',
    'total_votes_cast',
    'section24a_votes',
    'special_votes'
]

for col in int_cols:
    df2009_renamed[col] = pd.to_numeric(df2009_renamed[col], errors='coerce').astype('Int64')

# convert registered population to int
df2019_renamed['registered_population'] = df2019_renamed['registered_population'].astype('Int64')

#### d. Fix decimal percentages in vote_turnout_percent columns.

In [127]:
for year, df in {'2004': df2004_renamed, '2009': df2009_renamed, '2014': df2014_renamed}.items():
    mask = df['voter_turnout_percent'] > 1
    df.loc[mask, 'voter_turnout_percent'] = df.loc[mask, 'voter_turnout_percent'] / 100
    print(f"{year} - values normalised: {mask.sum()}")
    print(f"{year} - min: {df['voter_turnout_percent'].min()}, max: {df['voter_turnout_percent'].max()}\n")

2004 - values normalised: 611933
2004 - min: 0.0, max: 8.0174

2009 - values normalised: 854651
2009 - min: 0.0, max: 4.4804

2014 - values normalised: 14031
2014 - min: 0.0, max: 1.0



### 2. MDB Wards and Voting Stations datasets

In [128]:
# read and initialise tables
dfmdb = pd.read_sql("SELECT * FROM bronze.\"mdb_wards_2020\"", con=engine)
dfvotingstations = pd.read_sql("SELECT * FROM bronze.\"voting_stations\"", con=engine)

#### a. Naming conventions

In [129]:
dfmdb.head()

,FID,Province,Municipali,CAT_B,WardNo,District,DistrictCo,Date,WardID,WardLabel,Shape__Area,Shape__Length
0,1,Western Cape,Beaufort West Local Municipality,WC053,1,Central Karoo,DC5,11/24/2020 12:00:00 AM,10503001,WC053_1,4.389042e+09,436837.154786
1,2,Western Cape,Beaufort West Local Municipality,WC053,2,Central Karoo,DC5,11/24/2020 12:00:00 AM,10503002,WC053_2,1.162735e+10,849632.681502
2,3,Western Cape,Beaufort West Local Municipality,WC053,3,Central Karoo,DC5,11/24/2020 12:00:00 AM,10503003,WC053_3,5.470460e+06,11681.842953
3,4,Western Cape,Beaufort West Local Municipality,WC053,4,Central Karoo,DC5,11/24/2020 12:00:00 AM,10503004,WC053_4,1.157549e+07,16311.388967
4,5,Western Cape,Beaufort West Local Municipality,WC053,5,Central Karoo,DC5,11/24/2020 12:00:00 AM,10503005,WC053_5,3.116989e+06,10117.518920


In [130]:
dfmdb.columns

Index(['FID', 'Province', 'Municipali', 'CAT_B', 'WardNo', 'District',
       'DistrictCo', 'Date', 'WardID', 'WardLabel', 'Shape__Area',
       'Shape__Length'],
      dtype='str')

In [131]:
dfmdb_renamed = dfmdb.rename(columns={
    'FID'      : 'id',
    'Province'    : 'province',
    'Municipali'  : 'municipality',
    'CAT_B'       : 'cat_b',
    'WardNo'      : 'ward_number',
    'District'    : 'district_name',
    'DistrictCo'  : 'district_code',
    'Date'        : 'date',
    'WardID'      : 'ward_id',
    'WardLabel'   : 'ward_label',
    'Shape__Area' : 'shape_area',
    'Shape__Length': 'shape_length'
})



In [132]:
dfvotingstations.head()

,vd_number,api_response,ingestion_timestamp,status,municipality_id,municipality_code,municipality_name,province_id,province_name,ward,geometry,station_name,station_latitude,station_longitude,station_street_name,station_suburb,station_town,station_type,station_open_time,station_close_time
0,10840474,"{'Ward': '21201024', 'Status': 'OK', 'Geometry...",2026-03-12 12:32:17.805469,OK,1113,EC121,EC121 - Mbhashe,1,Eastern Cape,21201024,"{'type': 'MultiPolygon', 'coordinates': [[[[28...",LOWER GWADU JUNIOR SECONDARY SCHOOL,-32.220439,28.667648,LOWER GWADU VILLAGE,FORT MALAN A/A,WILLOWVALE,PERMANENT (P),,
1,10840485,"{'Ward': '21201024', 'Status': 'OK', 'Geometry...",2026-03-12 12:32:17.805469,OK,1113,EC121,EC121 - Mbhashe,1,Eastern Cape,21201024,"{'type': 'MultiPolygon', 'coordinates': [[[[28...",MSENGENI JUNIOR SECONDARY SCHOOL,-32.219779,28.626156,NKELEKETHE VILLAGE,NKELEKETHE A/A,WILLOWVALE,PERMANENT (P),,
2,10840496,"{'Ward': '21201023', 'Status': 'OK', 'Geometry...",2026-03-12 12:32:17.805469,OK,1113,EC121,EC121 - Mbhashe,1,Eastern Cape,21201023,"{'type': 'MultiPolygon', 'coordinates': [[[[28...",ZWELIDUMILE SENIOR SECONDARY SCHOOL,-32.298684,28.580454,LUVUNDU VILLAGE,NGOZI A/A,WILLOWVALE,PERMANENT (P),,
3,10840508,"{'Ward': '21201029', 'Status': 'OK', 'Geometry...",2026-03-12 12:32:17.805469,OK,1113,EC121,EC121 - Mbhashe,1,Eastern Cape,21201029,"{'type': 'MultiPolygon', 'coordinates': [[[[28...",NGWANE JUNIOR SECONDARY SCHOOL,-32.288227,28.749851,NGWANE VILLAGE,NQABARA A/A,WILLOWVALE,PERMANENT (P),,
4,10840519,"{'Ward': '21201023', 'Status': 'OK', 'Geometry...",2026-03-12 12:32:17.805469,OK,1113,EC121,EC121 - Mbhashe,1,Eastern Cape,21201023,"{'type': 'MultiPolygon', 'coordinates': [[[[28...",HADI JUNIOR SECONDARY SCHOOL,-32.413122,28.609463,QWANINGA VILLAGE,SEBELI A/A,WILLOWVALE,PERMANENT (P),,


#### b. Missing Values

In [133]:
dfmdb_renamed.isnull().sum()

id                 0
province           0
municipality       0
cat_b              0
ward_number        0
district_name      0
district_code      0
date             319
ward_id            0
ward_label         0
shape_area         0
shape_length       0
dtype: int64

We will drop the date column as it adds no analytical value.

In [134]:
dfmdb_renamed.drop(columns=['date'], inplace=True)

In [135]:
dfvotingstations.isnull().sum()

vd_number                0
api_response             0
ingestion_timestamp      0
status                   0
municipality_id          0
municipality_code        0
municipality_name        0
province_id              0
province_name            0
ward                     0
geometry               654
station_name           654
station_latitude       654
station_longitude      654
station_street_name    654
station_suburb         654
station_town           654
station_type           654
station_open_time      654
station_close_time     654
dtype: int64

We will drop the "station_open_time" and "station_close_time" columns as they add no value to our analytics. 

In [136]:
dfvotingstations.drop(columns=['station_open_time', 'station_close_time','api_response'], inplace=True)

#### c. Data Types

In [137]:
# Check datatypes for all years
for year, df in {'dfmdb_renamed': dfmdb_renamed, 'dfvotingstations':dfvotingstations}.items():
    print(f"\n{'='*50}")
    print(f" {year} NPE")
    print(f"{'='*50}")
    print(df.dtypes)


 dfmdb_renamed NPE
id                 int64
province             str
municipality         str
cat_b                str
ward_number        int64
district_name        str
district_code        str
ward_id            int64
ward_label           str
shape_area       float64
shape_length     float64
dtype: object

 dfvotingstations NPE
vd_number                       int64
ingestion_timestamp    datetime64[us]
status                            str
municipality_id                   str
municipality_code                 str
municipality_name                 str
province_id                       str
province_name                     str
ward                              str
geometry                       object
station_name                      str
station_latitude                  str
station_longitude                 str
station_street_name               str
station_suburb                    str
station_town                      str
station_type                      str
dtype: object


### 3. Demographic datasets

In [138]:
# read and initialise tables
dfage = pd.read_sql("SELECT * FROM bronze.\"pp_age_group_27_10_2025\"", con=engine)
dfgroup = pd.read_sql("SELECT * FROM bronze.\"pp_population_group_27_10_2025\"", con=engine)
dfsex = pd.read_sql("SELECT * FROM bronze.\"pp_sex_27_10_2025\"", con=engine)

#### a. Naming conventions

In [139]:

dfage_renamed = dfage.rename(columns={
    'Ward_Code': 'ward_number',
    '0-4'      : 'age_0_4',
    '5-9'      : 'age_5_9',
    '10-14'    : 'age_10_14',
    '15-19'    : 'age_15_19',
    '20-24'    : 'age_20_24',
    '25-29'    : 'age_25_29',
    '30-34'    : 'age_30_34',
    '35-39'    : 'age_35_39',
    '40-44'    : 'age_40_44',
    '45-49'    : 'age_45_49',
    '50-54'    : 'age_50_54',
    '55-59'    : 'age_55_59',
    '60-64'    : 'age_60_64',
    '65-69'    : 'age_65_69',
    '70-74'    : 'age_70_74',
    '75-79'    : 'age_75_79',
    '80-84'    : 'age_80_84',
    '85+'      : 'age_85_plus',
    'Total'    : 'total_population'
})

dfgroup_renamed = dfgroup.rename(columns={
    'Ward_Code'    : 'ward_number',
    'Black African': 'black_african',
    'Coloured'     : 'coloured',
    'Indian/Asian' : 'indian_or_asian',
    'White'        : 'white',
    'Other'        : 'other',
    'Total'        : 'total_population'
})

dfsex_renamed = dfsex.rename(columns={
    'WARD_CODE': 'ward_number',
    'Male'     : 'male',
    'Female'   : 'female',
    'Total'    : 'total_population'
})


df_population = (
    dfage_renamed
    .merge(dfgroup_renamed, on='ward_number', suffixes=('', '_groups'))
    .merge(dfsex_renamed,  on='ward_number', suffixes=('', '_sexes'))
)

# Drop duplicate total_population columns from merges
df_population = df_population.drop(columns=['total_population_groups', 'total_population_sexes'])

print(df_population.shape)
print(df_population.columns.tolist())

(4469, 27)
['ward_number', 'age_0_4', 'age_5_9', 'age_10_14', 'age_15_19', 'age_20_24', 'age_25_29', 'age_30_34', 'age_35_39', 'age_40_44', 'age_45_49', 'age_50_54', 'age_55_59', 'age_60_64', 'age_65_69', 'age_70_74', 'age_75_79', 'age_80_84', 'age_85_plus', 'total_population', 'black_african', 'coloured', 'indian_or_asian', 'white', 'other', 'male', 'female']


In [140]:
df_population.head()

,ward_number,age_0_4,age_5_9,age_10_14,age_15_19,age_20_24,age_25_29,age_30_34,age_35_39,age_40_44,...,age_80_84,age_85_plus,total_population,black_african,coloured,indian_or_asian,white,other,male,female
0,19100001,1362.206750,1328.730644,1483.628073,1241.239486,1775.578055,1612.088019,1877.091069,2381.348515,2474.858317,...,768.967086,652.045438,27755.734811,1397.159282,4030.389019,2591.734981,19132.407775,604.043754,13064.256842,14691.477969
1,19100002,1894.185815,1847.409981,2106.762805,1890.553938,2622.886505,2464.201139,2863.389243,3356.568785,3250.315266,...,624.834491,602.946737,35808.392693,4117.202067,11317.277440,1387.653977,17916.317711,1069.941498,17134.341705,18674.050988
2,19100003,1397.882218,1269.161700,1487.216323,1412.103176,1785.923849,1871.148246,2411.785265,2734.648507,2467.418485,...,418.876091,297.087852,27653.564529,2222.659174,5835.356993,511.707429,18412.511136,671.329797,13250.509612,14403.054916
3,19100004,2574.798817,2188.908451,2069.172671,1968.012803,2679.446069,3404.522642,3810.879001,3627.592348,2765.668554,...,121.672594,97.119741,32783.792978,17234.499447,4244.382291,1306.205198,9518.805089,479.900953,16253.568171,16530.224807
4,19100005,2118.205693,1689.117520,1943.022800,1780.135409,2553.700273,3886.041704,4367.624778,4064.289371,3183.182190,...,471.377675,314.997962,39645.384884,3603.590223,7426.404674,2130.491873,25738.503010,746.395104,19212.354124,20433.030760


In [141]:
print(df2004_renamed.columns)
print(df2009_renamed.columns)
print(df2014_renamed.columns)
print(df2019_renamed.columns)
print(df2024_renamed.columns)

Index(['electoral_event', 'province', 'municipality', 'vd_number',
       'party_name', 'registered_population', 'voter_turnout_percent',
       'valid_votes', 'spoilt_votes', 'total_votes_cast'],
      dtype='str')
Index(['electoral_event', 'province', 'municipality', 'ward_number',
       'vd_number', 'party_name', 'registered_population',
       'voter_turnout_percent', 'valid_votes', 'spoilt_votes',
       'total_votes_cast', 'section24a_votes', 'special_votes'],
      dtype='str')
Index(['electoral_event', 'province', 'municipality', 'ward_number',
       'vd_number', 'party_name', 'registered_population',
       'voter_turnout_percent', 'valid_votes', 'spoilt_votes',
       'total_votes_cast', 'section24a_votes', 'special_votes'],
      dtype='str')
Index(['province', 'municipality', 'vd_number', 'vs_name',
       'registered_population', 'spoilt_votes', 'valid_votes', 'party_name',
       'party_votes'],
      dtype='str')
Index(['province', 'municipality', 'vd_number', 'vs_name

In [142]:
# Your DataFrames + corresponding years
dfs = {
    2004: df2004_renamed,
    2009: df2009_renamed,
    2014: df2014_renamed,
    2019: df2019_renamed,
    2024: df2024_renamed
}

def standardize_basic(df, year):
    df = df.copy()


    # Add YEAR column

    df['year'] = year


    #  Ensure total_votes_cast exists
  
    if 'total_votes_cast' not in df.columns:
        df['total_votes_cast'] = (
            df.get('spoilt_votes', 0).fillna(0) +
            df.get('valid_votes', 0).fillna(0)
        )

    return df

# Apply to all datasets
dfs_standardized = {year: standardize_basic(df, year) for year, df in dfs.items()}

combined_df = pd.concat(
    dfs_standardized.values(),
    ignore_index=True,
    sort=False 
)

# Check result
print(combined_df[['year', 'total_votes_cast']].head())

   year  total_votes_cast
0  2004                 9
1  2004                 9
2  2004                 9
3  2004                 9
4  2004                 9


In [143]:
election_results_silver=combined_df
election_results_silver.head()

,electoral_event,province,municipality,vd_number,party_name,registered_population,voter_turnout_percent,valid_votes,spoilt_votes,total_votes_cast,year,ward_number,section24a_votes,special_votes,vs_name,party_votes,generated_datetime
0,14 APR 2004 NATIONAL ELECTION,EASTERN CAPE,UNKNOWN,10590006,AFRICAN CHRISTIAN DEMOCRATIC PARTY,0,0.0,0,1,9,2004,<NA>,<NA>,<NA>,NaN,NaN,NaN
1,14 APR 2004 NATIONAL ELECTION,EASTERN CAPE,UNKNOWN,10590006,AFRICAN NATIONAL CONGRESS,0,0.0,0,1,9,2004,<NA>,<NA>,<NA>,NaN,NaN,NaN
2,14 APR 2004 NATIONAL ELECTION,EASTERN CAPE,UNKNOWN,10590006,AZANIAN PEOPLE'S ORGANISATION,0,0.0,0,1,9,2004,<NA>,<NA>,<NA>,NaN,NaN,NaN
3,14 APR 2004 NATIONAL ELECTION,EASTERN CAPE,UNKNOWN,10590006,CHRISTIAN DEMOCRATIC PARTY,0,0.0,0,1,9,2004,<NA>,<NA>,<NA>,NaN,NaN,NaN
4,14 APR 2004 NATIONAL ELECTION,EASTERN CAPE,UNKNOWN,10590006,DEMOCRATIC ALLIANCE/DEMOKRATIESE ALLIANSIE,0,0.0,7,1,9,2004,<NA>,<NA>,<NA>,NaN,NaN,NaN


In [145]:
mdf_silver = dfmdb_renamed
voting_stations_silver = dfvotingstations
population_silver = df_population
election_results_silver = election_results_silver

import json
from sqlalchemy.dialects.postgresql import JSONB, DOUBLE_PRECISION

dfs_silver = {
    "mdb": mdf_silver,
    "voting_stations": voting_stations_silver,
    "population": population_silver,
    "election_results": election_results_silver
}

for table_name, df in dfs_silver.items():
    df = df.copy()  

    dtype_mapping = None

    # Handle geometry columns 
    if 'geometry' in df.columns:
        df['geometry'] = df['geometry'].apply(json.dumps)
        dtype_mapping = {"geometry": JSONB}

    # Handle derived metrics 
    if table_name == "election_results":
        dtype_mapping = {
            "voter_turnout_percent": DOUBLE_PRECISION,
        }

    # Save to silver schema in chunks
    df.to_sql(
        name=table_name,
        con=engine,
        schema="silver",
        if_exists="replace",
        index=False,
        dtype=dtype_mapping,
        chunksize=50000,  
        method='multi'   
    )

    print(f"Saved {table_name} to silver layer.")

Saved mdb to silver layer.
Saved voting_stations to silver layer.
Saved population to silver layer.
Saved election_results to silver layer.
